### PR TITLE
[REEF-940] Create .NET job resource uploader using IFileSystem

### PR DESCRIPTION
**Pull request by anupam128**
JIRA:
 [REEF-940](https://issues.apache.org/jira/browse/REEF-940)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            {
+                LastModificationUnixTimestamp = DateTimeToUnixTimestamp(fileStatus.ModificationTime),
+                RemoteUploadPath = remoteFileUri.AbsoluteUri,
+                ResourceSize = fileStatus.LengthBytes
+            };
+        }
+
+        private long DateTimeToUnixTimestamp(DateTime dateTime)
+        {
+            return (long) (dateTime - Epoch).TotalSeconds;
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: UTC?

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            _fileSystem = fileSystem;
+            _resourceArchiveFileGenerator = resourceArchiveFileGenerator;
+            _jobSubmissionDirectoryProvider = jobSubmissionDirectoryProvider;
+        }
+
+        public JobResource UploadJobResource(string driverLocalFolderPath)
+        {
+            driverLocalFolderPath = driverLocalFolderPath.TrimEnd('\\') + @"\";
+            var driverUploadPath = _jobSubmissionDirectoryProvider.GetJobSubmissionDirectory().TrimEnd('/') + @"/";
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: An orthogonal issue, but ``GetJobSubmissionDirectly()`` should probably specify whether it's local or remote in its naming. I'm always a bit confused.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            Log.Log(Level.Info, "DriverFolderPath: {0} DriverUploadPath: {1}", driverLocalFolderPath, driverUploadPath);
+            var archivePath = _resourceArchiveFileGenerator.CreateArchiveToUpload(driverLocalFolderPath);
+
+            var destinationPath = Path.Combine(driverUploadPath, Path.GetFileName(archivePath));
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: It seems like ``driverUploadPath`` uses front-slashes, while ``Path.Combine`` can use back-slashes, depending on the system. Have you confirmed the behavior?

**Comment by afchung** [comment]
@anupam128 This looks good in general, just a few minor comments. Thanks!

File: lang/cs/Org.Apache.REEF.Client.Tests/Org.Apache.REEF.Client.Tests.csproj [hunk 4]
```diff
@@ -59,6 +60,10 @@ under the License.
       <Project>{5094c35b-4fdb-4322-ac05-45d684501cbf}</Project>
       <Name>Org.Apache.REEF.Client</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Org.Apache.REEF.IO\Org.Apache.REEF.IO.csproj">
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client.Tests/Org.Apache.REEF.Client.Tests.csproj]: We usually replace the `..` with a variable.

File: lang/cs/Org.Apache.REEF.Client/Org.Apache.REEF.Client.csproj [hunk 5]
```diff
@@ -177,6 +178,10 @@ under the License.
       <Project>{75503f90-7b82-4762-9997-94b5c68f15db}</Project>
       <Name>Org.Apache.REEF.Examples</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Org.Apache.REEF.IO\Org.Apache.REEF.IO.csproj">
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Org.Apache.REEF.Client.csproj]: We usually replace the `..` with a variable.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+using System.IO;
+using Org.Apache.REEF.Client.Common;
+using Org.Apache.REEF.Client.Yarn;
+using Org.Apache.REEF.IO.FileSystem;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Client.YARN.RestClient
+{
+    public sealed class JobResourceUploader : IJobResourceUploader
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: Does this have to be `public`? Also, please document the class and consider renaming it to something more explicit, e.g. `FileSystemJibUploader` or such.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            _fileSystem = fileSystem;
+            _resourceArchiveFileGenerator = resourceArchiveFileGenerator;
+            _jobSubmissionDirectoryProvider = jobSubmissionDirectoryProvider;
+        }
+
+        public JobResource UploadJobResource(string driverLocalFolderPath)
+        {
+            driverLocalFolderPath = driverLocalFolderPath.TrimEnd('\\') + @"\";
+            var driverUploadPath = _jobSubmissionDirectoryProvider.GetJobSubmissionDirectory().TrimEnd('/') + @"/";
+            Log.Log(Level.Info, "DriverFolderPath: {0} DriverUploadPath: {1}", driverLocalFolderPath, driverUploadPath);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: This should likely be logged to the `Fine` level.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            var archivePath = _resourceArchiveFileGenerator.CreateArchiveToUpload(driverLocalFolderPath);
+
+            var destinationPath = Path.Combine(driverUploadPath, Path.GetFileName(archivePath));
+            var remoteFileUri = _fileSystem.CreateUriForPath(destinationPath);
+            Log.Log(Level.Info, @"Copy {0} to {1}", archivePath, remoteFileUri);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: Same as above: Log to `Fine`

**Comment by markusweimer** [comment]
@anupam128 I've done a pass.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs (hunk 6 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: This is not the default implementation for `IJobResourceUploader`. The user might have to explicitly bind this implementation in their client. I am open to making it internal for now.

**Comment by anupam128** [comment]
@afchung @markusweimer I have made the changes.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs (hunk 6 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/JobResourceUploader.cs]: We could expose it via a `ConfigurationModule` only, like we do for file systems?

**Comment by afchung** [comment]
Looks good. Have you done any E2E tests?

**Comment by anupam128** [comment]
@afchung I don't have a way to do E2E test yet (HadoopFileSystem does not have GetFileStatus API)

REEF-922 with include this E2E testing.

**Comment by afchung** [comment]
@anupam128 OK, I'll test and merge after @markusweimer signs off.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/FileSystemJobResourceUploader.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+namespace Org.Apache.REEF.Client.YARN.RestClient
+{
+    /// <summary>
+    /// Provides FileSystem agnostic job resource uploader.
+    /// User can provide custome implementation of 
+    /// <see cref="IFileSystem"/> for their choice of DFS.
+    /// </summary>
+    internal sealed class FileSystemJobResourceUploader : IJobResourceUploader
+    {
+        private static readonly Logger Log = Logger.GetLogger(typeof(LegacyJobResourceUploader));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/FileSystemJobResourceUploader.cs]: This uses the wrong logger name, I believe?

**Comment by markusweimer** [comment]
I did another pass, left a comment. Other than that: +1 on merging.

**Comment by anupam128** [comment]
@afchung @markusweimer thanks for reviewing. I have made the change.

**Comment by afchung** [comment]
I'll test and merge.

**Pull request closed without merging**